pkg/ring/lifecycler.go

```diff
@@ -456,6 +477,17 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 
 		ingesterDesc, ok := ringDesc.Ingesters[i.ID]
 		if !ok {
+			// We use the tokens from the file only if it does not exist in the ring yet.
+			if len(tokensFromFile) > 0 {
+				level.Info(util.Logger).Log("msg", "adding tokens from file", "num_tokens", len(tokensFromFile))
+				if len(tokensFromFile) == i.cfg.NumTokens {
+					i.setState(ACTIVE)
```

Comment:
what exactly happens if there isn't enough of them? Ingester stays in PENDING state, and then what?

Reply:
The `autoJoin` takes care of adding more tokenshttps://github.com/cortexproject/cortex/blob/41d37d8f58c8ca5ba0ea44666fcac8e2a90e5290/pkg/ring/lifecycler.go#L580-L615

Now this makes me wonder what will happen if we have more tokens in the file than `i.cfg.NumTokens` - that is not handled yet. Should we ignore the file completely in that case or take a subset of it?

Reply:
> The `autoJoin` takes care of adding more tokenshttps://github.com/cortexproject/cortex/blob/41d37d8f58c8ca5ba0ea44666fcac8e2a90e5290/pkg/ring/lifecycler.go#L580-L615

Ah, so fixing the "old bug" gets more important :)

> Now this makes me wonder what will happen if we have more tokens in the file than `i.cfg.NumTokens` - that is not handled yet. Should we ignore the file completely in that case or take a subset of it?

This is part of WAL work... would it be safe to take subset of tokens only in that context?

Reply:
Yup, should be fine, will update accordingly